File: binary_search/BinarySearch.java
import java.util.Arrays;
 
public class BinarySearch{
    public static int binary_search_iterative(int[] arr, int item){
        int low = 0;
        int high = arr.length - 1; // length of the array gives the last index as the highest

        while(low <= high){ // as long as low and high don't overlap
            int mid = (low + high) / 2;
            if (arr[mid] == item){
                return mid; // if guess is correct
            }

            if (item > arr[mid]){
                low = mid + 1;
            }

            else{
                high = mid - 1;
            }
        }
        return -1;

    }

    public static void main(String args[]){
        int array [] = {3, 1, 5, 2, 6, 9};
        Arrays.sort(array); // sorts in ascending order
        System.out.println(binary_search_iterative(array, 5));
        System.out.print(binary_search_iterative(array, 10));
    }

}

File: binary_search/binary_search.py
def binary_search(arr, item):
    high = len(arr) - 1 #last index in the list
    low = 0 # first index in the list

    while low <= high:
        mid = (high + low) // 2 # returns a rounded off number
        guess = arr[mid] # gets the actual element
        if guess == item:
            return mid # returns index of the element

        if item > guess: # item is greater than the mid
            low = mid + 1 # discards every element lower than mid including mid

        else:
            high = mid - 1 # discards every element higher than mid including mid

    return None # item is not found in the list

list = [4, 5, 6, 7, 8, 3, 2, 9]
sort_list = sorted(list)
print(binary_search(sort_list, 7))
print(binary_search(sort_list, -1))
